lib/rules/no-restricted-properties.js

```diff
@@ -51,38 +68,96 @@ module.exports = {
 
-            restrictions.get(objectName).set(propertyName, {
-                message: option.message
-            });
+            if (matchedObjectProperty) {
+                const message = matchedObjectProperty.message ? " " + matchedObjectProperty.message : "";
 
-            return restrictions;
-        }, new Map());
+                context.report(node, "'{{objectName}}.{{propertyName}}' is restricted from being used.{{message}}", {
+                    objectName,
+                    propertyName,
+                    message
+                });
+            } else if (globalMatchedProperty) {
+                const message = globalMatchedProperty.message ? " " + globalMatchedProperty.message : "";
+
+                context.report(node, "'{{propertyName}}' is restricted from being used.{{message}}", {
+                    propertyName,
+                    message
+                });
+            }
+        }
+
+        /**
+        * Checks property accesses in a destructuring assignment expression, e.g. `var foo; ({foo} = bar);`
+        * @param {ASTNode} node An AssignmentExpression or AssignmentPattern node
+        * @returns {undefined}
+        */
+        function checkDestructuringAssignment(node) {
+            if (node.right.type === "Identifier") {
+                const objectName = node.right.name;
+
+                if (node.left.type === "ObjectPattern") {
+                    node.left.properties.forEach(property => {
+                        checkPropertyAccess(node.left, objectName, property.key.name || property.key.value);
+                    });
+                }
+            }
+        }
 
         return {
             MemberExpression(node) {
                 const objectName = node.object && node.object.name;
                 const propertyName = astUtils.getStaticPropertyName(node);
-                const matchedObject = restrictedProperties.get(objectName);
-                const matchedObjectProperty = matchedObject && matchedObject.get(propertyName);
 
-                if (matchedObjectProperty) {
-                    const message = matchedObjectProperty.message ? " " + matchedObjectProperty.message : "";
+                checkPropertyAccess(node, objectName, propertyName);
+            },
+            VariableDeclarator(node) {
+                if (node.init && node.init.type === "Identifier") {
+                    const objectName = node.init.name;
 
-                    context.report(node, "'{{objectName}}.{{propertyName}}' is restricted from being used.{{message}}", {
-                        objectName,
-                        propertyName,
-                        message
-                    });
+                    if (node.id.type === "ObjectPattern") {
+                        node.id.properties.forEach(property => {
+                            checkPropertyAccess(node.id, objectName, property.key.name || property.key.value);
```

Comment:
This should ignore the property if the property has dynamic name.

```js
const name = astUtils.getStaticPropertyName(property);
if (name) {
    checkPropertyAccess(node.id, objectName, name);
}
```

There is the same code in `checkDestructuringAssignment` function.